src/Umbraco.Web/PublishedCache/NuCache/DataSource/MsgPackContentNestedDataSerializer.cs

```diff
@@ -0,0 +1,199 @@
+
+        public MsgPackContentNestedDataSerializer(IPropertyCompressionOptions propertyOptions)
+        {
+            _propertyOptions = propertyOptions ?? throw new ArgumentNullException(nameof(propertyOptions));
+
+            var defaultOptions = ContractlessStandardResolver.Options;
+            var resolver = CompositeResolver.Create(
+
+                // TODO: We want to be able to intern the strings for aliases when deserializing like we do for Newtonsoft but I'm unsure exactly how
+                // to do that but it would seem to be with a custom message pack resolver but I haven't quite figured out based on the docs how
+                // to do that since that is part of the int key -> string mapping operation, might have to see the source code to figure that one out.
+
+                // resolver custom types first
+                // new ContentNestedDataResolver(),
+
+                // finally use standard resolver
+                defaultOptions.Resolver
+            );
+
+            _options = defaultOptions
+                .WithResolver(resolver)
+                .WithCompression(MessagePackCompression.Lz4BlockArray);            
+        }
+
+        public string ToJson(byte[] bin)
+        {
+            var json = MessagePackSerializer.ConvertToJson(bin, _options);
+            return json;
+        }
+
+        public ContentCacheDataModel Deserialize(int contentTypeId, string stringData, byte[] byteData)
+        {
+            if (stringData != null)
+            {
+                // NOTE: We don't really support strings but it's possible if manually used (i.e. tests)
+                var bin = Convert.FromBase64String(stringData);
+                var content = MessagePackSerializer.Deserialize<ContentCacheDataModel>(bin, _options);
+                Expand(contentTypeId, content);
+                return content;
+            }
+            else if (byteData != null)
+            {
+                var content = MessagePackSerializer.Deserialize<ContentCacheDataModel>(byteData, _options);
+                Expand(contentTypeId, content);
+                return content;
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        public ContentCacheDataSerializationResult Serialize(int contentTypeId, ContentCacheDataModel model)
+        {
+            Compress(contentTypeId, model);
+            var bytes = MessagePackSerializer.Serialize(model, _options);
+            return new ContentCacheDataSerializationResult(null, bytes);
+        }
+
+        /// <summary>
+        /// Used during serialization to compress properties
+        /// </summary>
+        /// <param name="model"></param>
+        private void Compress(int contentTypeId, ContentCacheDataModel model)
+        {
+            foreach(var propertyAliasToData in model.PropertyData)
+            {
+                if (_propertyOptions.IsCompressed(contentTypeId, propertyAliasToData.Key))
```

Comment:
I don't think this should be default as it's quite slow on decompression

Reply:
I'm assuming you mean the double compression as the default? Right now if you want to test the difference, you can replace `IContentCacheDataSerializerFactory` in DI with an overridden version of `MsgPackContentNestedDataSerializerFactory` and just return a new instance of `MsgPackContentNestedDataSerializer` with empty `IPropertyCompressionOptions` that always returns false. 

Is it just the first published query that is slow since it needs to decompress things? Since once the first one is done it will be decompressed in memory right?

Reply:
If you have a few different queries it's slow for a while

Reply:
OK cool, will change how this works with the ability to opt-in to manage what might be compressed in memory but will leave it uncompressed (so we only compress with MsgPack at the db level). This task is still ongoing, sorry the work is slow but I can only allocate a bit of time each sprint.

Reply:
Perharps we could break this up into multiple smaller prs.

Reply:
Yes that could be possible, I've just finished re-reviewing a lot of it and making some notes. I need to resolve this one https://github.com/umbraco/Umbraco-CMS/pull/9440/files since this https://github.com/umbraco/Umbraco-CMS/pull/9440#discussion_r544006996 isn't quite right and have run into issues there. 

* Compression is one thing - which is a majority of these changes
* Improved Sql indexes + INCLUDE support could be separated 
* Improved Sql queries could prob be separated out but that does have some overlap with the compression db changes
* Some of the interning could be separated out if it's easy but there's a lot of stuff for that which relates to the changes for compression
* Other "Fixes a startup issue that doesn't bubble the underlying boot failed exception if the container fails." could be separated

If you have time let me know but I will continue to resolve the last bits here regarding compression